lib/bolt_server/transport_app.rb

```diff
@@ -135,11 +136,30 @@ def task_helper(target, task, parameters)
         end
       end
 
-      @executor.run_task(target, task, parameters).each do |result|
-        value = result.value
-        next unless value.is_a?(Hash)
-        next unless value.key?('_sensitive')
-        value['_sensitive'] = value['_sensitive'].unwrap
+      if timeout
+        task_future = Concurrent::Future.new do
+          @executor.run_task(target, task, parameters).each do |result|
+            value = result.value
+            next unless value.is_a?(Hash)
+            next unless value.key?('_sensitive')
+            value['_sensitive'] = value['_sensitive'].unwrap
+          end
+        end
+        task_future.execute
+        task_complete = task_future.wait_or_cancel(timeout)
```

Comment:
The `wait_or_cancel` method does not actually do what I thought. The cancel only happens if the future does not actually begin, in the case of a timeout the future is not properly resolved. This is showing up in the rspec tests hanging in CI. I think I will need to use something else for this:
https://github.com/ruby-concurrency/concurrent-ruby/blob/master/docs-source/cancellation.in.md (see `Run only for limited time – Timeout replacement` section).

Reply:
UPDATE: given we are pinned back to an old concurrent ruby version I'm just going to do this with simple build in ruby threads.